### PR TITLE
fix: yield waited responses in the order that they occured

### LINKED ISSUE #27337: `cy.wait` on an array of the same alias does not resolve in expected order
### Current behavior

`cy.wait` accepts an array of aliases to await - this behavior works as expected when different aliases as supplied (`cy.wait(['@a', '@b'])`), however when the same alias is supplied multiple times the resolved subjects are not ordered from first-to-last as one would expect.

For example:

```js
cy.intercept('url').as('@req')

//...multiple calls are made to the intercepted endpoint...

cy.wait(['@req', '@req', '@req']).then((subjects) => {
    // I would expect `subjects` to be in order of incidence (first intercepted request, second, third...)
    // order of `subjects` appears to be arbitrary
	// subjects[0] does not always correspond to the first interception
	// ditto with subjects[1], [2], etc
})
```

### Desired behavior

Users can currently await serially to get items in order:

```js
cy.wait('@req').then((req) => ...do your assertions on the first interception...)
cy.wait('@req').then((req) => ...do your assertions on the second interception...)
cy.wait('@req').then((req) => ...do your assertions on the third interception...)
```

but this prevents easy comparison of multiple awaited subjects. One workaround that been identified is to wait serially and combine into an array, but this relies on variables and state outside of the Cypress command chain which has inherent risks:

```js
const waitedRequests = []
for (let i = 0; i < 3; i++) {
  cy.wait('@req').then((req) => {
    waitedRequests.push(req)
  })
}

cy.then(() => {
 // `waitedRequests` should contain all three requests in order here
})
```

### Test code to reproduce

https://github.com/mike-plummer/cypress-wait-alias-array

### Cypress Version

12.15.0

### Node version

16.17.1

### Operating System

macOS 13.4.1

### Debug Logs

_No response_

### Other

Based off of cypress-io/cypress-support-internal/issues/13

### PR DESCRIPTION
**Pull request by astone123**
- Closes https://github.com/cypress-io/cypress/issues/27337

### Additional details

This PR sorts the yielded responses from a `cy.wait` command by `browserRequestId` so that the responses are returned in the correct order even when the same alias name is used.

### Steps to test

Follow the reproduction steps in the original issue

### PR Tasks

- [ ] Have tests been added/updated?
- [ ] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [ ] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

### CODE REVIEW

**Review by dkasper-was-taken — DISMISSED** [review]

**Review by chrisbreiding — CHANGES_REQUESTED** [review]
We should add a test for this.

**Comment by dkasper-was-taken** [comment]
> We should add a test for this.

@chrisbreiding Test case added.

File: packages/driver/cypress/e2e/commands/waiting.cy.js [hunk 1]
```diff
@@ -737,6 +737,44 @@ describe('src/cy/commands/waiting', () => {
...
+          xhrGet(win, '/epsilon')
+          xhrGet(win, '/beta')
+          xhrGet(win, '/gamma')
+          xhrGet(win, '/delta')
+          xhrGet(win, '/alpha')
+
+          return null
+        })
+
+        cy.wait(['@getAlpha', '@getBeta', '@getGamma', '@getDelta', '@getEpsilon']).then((responses) => {
```
Comment (chrisbreiding) [inline packages/driver/cypress/e2e/commands/waiting.cy.js:787]: The [original issue](https://github.com/cypress-io/cypress/issues/27337) concerns using the same alias for a request that's called multiple times, but this is testing different aliases. Seems like this test should set up one alias on a particular request interception, trigger that request multiple times, and then assert that the responses are in the right order.

**Review by chrisbreiding — CHANGES_REQUESTED** [review]

File: packages/driver/cypress/e2e/commands/waiting.cy.js (hunk 1 above)
Comment (dkasper-was-taken) [reply packages/driver/cypress/e2e/commands/waiting.cy.js:787]: Updated with new test.

Comment (chrisbreiding) [reply packages/driver/cypress/e2e/commands/waiting.cy.js:787]: Now that I think about _this_ test some more ... is this actually the behavior we want? It seems counter-intuitive for the user to pass in the aliases in a specific order and then the responses are provided in a different order. I think the ordering may need to be more nuanced so it only affects aliases that are the same and not aliases that are different. So if the requests are:

```
alpha (1)
alpha (2)
beta
alpha (3)
gamma (1)
delta
gamma (2)
```

And the user's test has:

```js
cy.wait(['@getGamma', '@getBeta', '@getAlpha', '@getGamma', '@getDelta', '@getAlpha', '@getAlpha'])
```

The responses should be in the following order:

```js
.then((responses) => {
  responses[0] // => gamma (1)
  responses[1] // => beta
  responses[2] // => alpha (1)
  responses[3] // => gamma (2)
  responses[4] // => delta
  responses[5] // => alpha (2)
  responses[6] // => alpha (3)
})
```

Does that make sense?

Comment (dkasper-was-taken) [reply packages/driver/cypress/e2e/commands/waiting.cy.js:787]: Yeah that makes sense, I'll take a crack at it. I wrote a version of that test and saw the behavior lacking. Good Call!

Comment (dkasper-was-taken) [reply packages/driver/cypress/e2e/commands/waiting.cy.js:768]: Updated to match.

File: packages/driver/src/cy/commands/waiting.ts [hunk 2]
```diff
@@ -250,7 +250,22 @@ export default (Commands, Cypress, cy, state) => {
     .then((responses) => {
       // if we only asked to wait for one alias
       // then return that, else return the array of xhr responses
-      const ret = responses.length === 1 ? responses[0] : responses
+      const ret = responses.length === 1 ? responses[0] : ((resp) => {
+        const respMap = new Map()
+        const respSeq = resp.map((r) => r.routeId)
+
+        resp.sort((a, b) => {
```
Comment (chrisbreiding) [inline packages/driver/src/cy/commands/waiting.ts:259]:
```suggestion
        // responses are sorted in the order the user specified them. if there are multiples of the same
        // alias awaited, they're sorted in execution order
        resp.sort((a, b) => {
```

**Review by chrisbreiding — CHANGES_REQUESTED** [review]
Over all looks good, though the new test seems to be [failing or flaky in webkit](https://cloud.cypress.io/projects/ypt4pf/runs/51408/test-results?actions=%5B%5D&browsers=%5B%5D&groups=%5B%5D&isFlaky=%5B%5D&modificationDateRange=%7B%22startDate%22%3A%221970-01-01%22%2C%22endDate%22%3A%222038-01-19%22%7D&orderBy=EXECUTION_ORDER&oses=%5B%5D&specs=%5B%5D&statuses=%5B%7B%22value%22%3A%22FAILED%22%2C%22label%22%3A%22FAILED%22%7D%5D&testingTypesEnum=%5B%5D&utm_source=github) 😔

File: packages/driver/src/cy/commands/waiting.ts (hunk 2 above)
Comment (dkasper-was-taken) [reply packages/driver/src/cy/commands/waiting.ts:259]: updated

**Comment by dkasper-was-taken** [comment]
> Over all looks good, though the new test seems to be [failing or flaky in webkit](https://cloud.cypress.io/projects/ypt4pf/runs/51408/test-results?actions=%5B%5D&browsers=%5B%5D&groups=%5B%5D&isFlaky=%5B%5D&modificationDateRange=%7B%22startDate%22%3A%221970-01-01%22%2C%22endDate%22%3A%222038-01-19%22%7D&orderBy=EXECUTION_ORDER&oses=%5B%5D&specs=%5B%5D&statuses=%5B%7B%22value%22%3A%22FAILED%22%2C%22label%22%3A%22FAILED%22%7D%5D&testingTypesEnum=%5B%5D&utm_source=github) 😔

@chrisbreiding Looking at webkit, it fails locally as well. Apparently if we `wait` 6 or more aliases the call fails to resolve regardless of timeout set. This occurs even if I revert the newly added sort logic.

Are there any known limitations with webkit around the # of xhr calls?

**Comment by chrisbreiding** [comment]
> Are there any known limitations with webkit around the # of xhr calls?

Not that I know of, but webkit has a lot of known and unknown issues, so it's not too surprising there's a weird issue with it.

**Comment by chrisbreiding** [comment]
I guess we should just have the test use less aliases so that it can work in webkit too.

**Review by chrisbreiding — APPROVED** [review]

File: packages/driver/cypress/e2e/commands/waiting.cy.js [hunk 3]
```diff
@@ -737,6 +737,83 @@ describe('src/cy/commands/waiting', () => {
       })
     })
 
+    describe('multiple alias execution order', () => {
+      beforeEach(() => {
+        cy.visit('/fixtures/jquery.html')
+      })
+
+      it('all responses returned in execution order - different aliases', () => {
```
Comment (mschile) [inline packages/driver/cypress/e2e/commands/waiting.cy.js]: Aren't the responses being returned in 'requested wait order', not 'execution order'?

File: packages/driver/cypress/e2e/commands/waiting.cy.js [hunk 4]
```diff
@@ -737,6 +737,83 @@ describe('src/cy/commands/waiting', () => {
       })
     })
 
+    describe('multiple alias execution order', () => {
```
Comment (mschile) [inline packages/driver/cypress/e2e/commands/waiting.cy.js]: Suggest removing this `describe` block and just moving the tests under `describe('multiple alias arguments',`

Comment (dkasper-was-taken) [reply packages/driver/cypress/e2e/commands/waiting.cy.js]: Updated.

File: packages/driver/cypress/e2e/commands/waiting.cy.js (hunk 3 above)
Comment (dkasper-was-taken) [reply packages/driver/cypress/e2e/commands/waiting.cy.js]: Renamed tests to avoid usage of "execution order'

**Review by mschile — APPROVED** [review]

**Pull request merged**